typegen/unification/subtyping.py

```diff
@@ -0,0 +1,156 @@
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.stdlib_path=}"
+            )
+        if self.proj_path is None:
+            raise AttributeError(
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.proj_path=}"
+            )
+        if self.venv_path is None:
+            raise AttributeError(
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.venv_path=}"
+            )
+        if self.only_replace_if_base_was_traced is None:
+            raise AttributeError(
+                f"{ReplaceSubTypesFilter.__name__} was not initialised properly: {self.only_replace_if_base_was_traced=}"
+            )
+
+        self._resolver = Resolver(self.stdlib_path, self.proj_path, self.venv_path)
+
+        subset = list(constants.TraceData.SCHEMA.keys())
+        subset.remove(constants.TraceData.VARTYPE_MODULE)
+        subset.remove(constants.TraceData.VARTYPE)
+        subset.remove(constants.TraceData.FILENAME)
+        grouped_trace_data = trace_data.groupby(subset, dropna=False)
+        processed_trace_data = grouped_trace_data.apply(
+            lambda group: self._update_group(trace_data, group)
+        )
+        typed = processed_trace_data.reset_index(drop=True).astype(
+            constants.TraceData.SCHEMA
+        )
+        typed.columns = constants.TraceData.SCHEMA.keys()
+        return typed
+
+    def _update_group(self, entire: pd.DataFrame, group):
+        modules_with_types_in_group = group[
+            [
+                constants.TraceData.FILENAME,
+                constants.TraceData.VARTYPE_MODULE,
+                constants.TraceData.VARTYPE,
+            ]
+        ]
+        common = self._get_common_base_type(modules_with_types_in_group)
+        if common is None:
+            return group
+
+        basetype_module, basetype = common
+        if self.only_replace_if_base_was_traced:
+            if basetype_module not in entire[constants.TraceData.VARTYPE_MODULE].values:
+                logger.debug(f"Discarding {common}; module was not found in trace data")
+                return group
+
+            if basetype not in entire[constants.TraceData.VARTYPE].values:
+                logger.debug(f"Discarding {common}; type was not found in trace data")
+                return group
+
+        group[constants.TraceData.VARTYPE_MODULE] = basetype_module
+        group[constants.TraceData.VARTYPE] = basetype
+        return group
+
+    def _get_common_base_type(
+        self, modules_with_types: pd.DataFrame
+    ) -> tuple[str | None, str] | None:
+        type2bases = {}
+        for _, row in modules_with_types.iterrows():
+            varmodule, vartyp = (
+                row[constants.TraceData.VARTYPE_MODULE],
+                row[constants.TraceData.VARTYPE],
+            )
+            types_topologically_sorted = self._get_type_and_mro(varmodule, vartyp)
+
+            # drop mros that are useless
```

Comment:
\# drop base types which are considered too common (ABC, object)

Also, use a "constant" list which contains the string names of the classes we do not want to have as a common base type.

Reply:
Done